packages/cli/src/lib/deploy/Deployer.ts

```diff
@@ -0,0 +1,229 @@
+  };
+  public manifest: DeployManifest;
+  private _config: DeployerConfig;
+
+  private constructor(
+    manifest: DeployManifest,
+    cache: CacheDirectory,
+    logger: Logger
+  ) {
+    this._config = {
+      logger,
+      cache,
+      defaultModulesCached: false,
+    };
+    this.manifest = manifest;
+  }
+
+  public static async create(
+    manifest: string,
+    logger: Logger
+  ): Promise<Deployer> {
+    const deployManifest = await loadDeployManifest(manifest, logger);
+    const cache = new CacheDirectory({
+      rootDir: path.dirname(manifest),
+      subDir: Deployer.cacheLayout.root,
+    });
+    return new Deployer(deployManifest, cache, logger);
+  }
+
+  public async run(): Promise<DeployJobResult[]> {
+    const allStepsFromAllJobs = Object.entries(this.manifest.jobs).flatMap(
+      ([jobName, job]) => {
+        return job.steps.map((step) => ({
+          jobName,
+          ...step,
+        }));
+      }
+    );
+
+    const packageNames = [
+      ...new Set(allStepsFromAllJobs.map((step) => step.package)),
+    ];
+
+    this._sanitizePackages(packageNames);
+
+    await this._cacheDeployModules(packageNames);
+
+    const packageMapEntries = await Promise.all(
+      packageNames.map(async (packageName) => {
+        const deployerPackage = await this._getDeployModule(packageName);
+        return [packageName, deployerPackage];
+      })
+    );
+
+    const packageMap = Object.fromEntries(packageMapEntries);
+
+    const stepToPackageMap: Record<
+      string,
+      DeployPackage & { jobName: string }
+    > = {};
+
+    for (const step of allStepsFromAllJobs) {
+      stepToPackageMap[step.name] = {
+        ...packageMap[step.package],
+        jobName: step.jobName,
+      };
+    }
+
+    this._validateManifestWithExts(this.manifest, stepToPackageMap);
+
+    const jobs = Object.entries(this.manifest.jobs).map(([jobName, job]) => {
+      const steps: DeployStep[] = job.steps.map((step) => {
+        return new DeployStep({
+          name: step.name,
+          uriOrStepResult: step.uri,
+          deployModule: stepToPackageMap[step.name].deployModule,
+          config: step.config ?? {},
+        });
+      });
+
+      return new DeployJob({
+        name: jobName,
+        steps,
+        config: job.config ?? {},
+        logger: this._config.logger,
+      });
+    });
+
+    return await Promise.all(jobs.map((job) => job.run()));
+  }
+
+  private _sanitizePackages(packages: string[]) {
+    const unrecognizedPackages: string[] = [];
+
+    const availableDeployers = fs.readdirSync(
+      path.join(__dirname, "..", "defaults", "deploy-modules")
+    );
+
+    packages.forEach((p) => {
+      if (!availableDeployers.includes(p)) {
+        unrecognizedPackages.push(p);
+      }
+    });
+
+    if (unrecognizedPackages.length) {
+      throw new Error(
+        `Unrecognized packages: ${unrecognizedPackages.join(", ")}`
+      );
+    }
+  }
+
+  private async _getDeployModule(
```

Comment:
rename: `getDeployPackage`

Reply:
Private properties & methods should be prefixed with an underscore. We do this in other places within the toolchain:
https://github.com/polywrap/toolchain/search?q=private+_

Reply:
It's not about `_` I meant renaming module -> package 😅